storm-core/src/jvm/org/apache/storm/daemon/worker/LogConfigManager.java

```diff
@@ -0,0 +1,154 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.storm.daemon.worker;
+
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.core.LoggerContext;
+import org.apache.logging.log4j.core.config.Configuration;
+import org.apache.logging.log4j.core.config.LoggerConfig;
+import org.apache.storm.generated.LogConfig;
+import org.apache.storm.generated.LogLevel;
+import org.apache.storm.generated.LogLevelAction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class LogConfigManager {
+
+    private static final Logger LOG = LoggerFactory.getLogger(LogConfigManager.class);
+
+    private final AtomicReference<TreeMap<String, LogLevel>> latestLogConfig;
+    private final Map<String, Level> originalLogLevels;
+
+    public LogConfigManager() {
+        this(new AtomicReference<>(new TreeMap<>()));
+    }
+
+    public LogConfigManager(AtomicReference<TreeMap<String, LogLevel>> latestLogConfig) {
+        this.latestLogConfig = latestLogConfig;
+        this.originalLogLevels = getLoggerLevels();
+        LOG.info("Started with log levels: {}", originalLogLevels);
+    }
+
+    public void processLogConfigChange(LogConfig logConfig) {
+        if (null != logConfig) {
+            LOG.debug("Processing received log config: {}", logConfig);
+            TreeMap<String, LogLevel> loggers = new TreeMap<>(logConfig.get_named_logger_level());
+            LoggerContext logContext = (LoggerContext) LogManager.getContext();
```

Comment:
In the previous code it passed in a false.  Not sure if it makes much difference in our case though.

https://logging.apache.org/log4j/2.x/log4j-api/apidocs/org/apache/logging/log4j/LogManager.html#getContext(boolean)